16.0 mig commown cooperative campaign

fcayre opened the pull request.

fcayre committed:
Add commown_cooperative_campaign module

fcayre committed:
Add optin at contract start

The first time a contract invoice is generated, an optin creation
request to the cooperative campaign WS is performed.

fcayre committed:
Add optout at contract end

fcayre committed:
Compute customer keys with a salted hash so that the WS db does not store clear phone numbers

We use keychain to store the salt in the database.

Note that the WS use HTTP Basic auth but the credentials are not stored
in the db, as we use request which can use a .netrc file.

fcayre committed:
Synchronize URLs with the real telecommown web service

fcayre committed:
Add a script to populate the WS database with current customers

fcayre committed:
Improve logging

fcayre committed:
Add helper to determine fully coop-campaign registered partners

fcayre committed:
Fix crash in raise

fcayre committed:
Remove wrong assertion

Our subscribers are not necessarily present in important-events: if they have not subscribe the other members' service.

fcayre committed:
Do not crash when a customer is already registered

The customer may have more than one contract.

fcayre committed:
Do not generate an error in case no identifier could be built for a contract

fcayre committed:
Improve optin http error handling

Double optins are detected and do not raise an error.
Other HTTP errors raise and stop the invoice procedure.

fcayre committed:
Invoice date and optin date equality implies a discount

fcayre committed:
Bump version to 1.0.0

fcayre committed:
Move partner identifier computation as a coupon campaign method

fcayre committed:
Coding-style improvements (silences flake8 warnings)

fcayre committed:
Bump version to 1.1.0

fcayre committed:
Do not call coop campaign WS on contract date_end change but on update

fcayre committed:
Improve logging to help debugging

fcayre committed:
Use queue_job to move the optout WS call into a job

This way, if the call crashes, the contract management team is not
blocked.

fcayre committed:
Bump version to 1.2.0

fcayre committed:
Test if contract partner has an identifier for the cooperative campaign before call optout service at end of contract

Also move some DB operations before the job submission to simplify the
job and improve performance.

fcayre committed:
Bump version to 1.2.1

fcayre committed:
Rename coupon campaign field and inverse its meaning

Old field was coupons_are_exclusive and was turned into its inverse can_cumulate.

fcayre committed:
Blind 12.0 port

fcayre committed:
Replace keychain module usage by an invisible attribute

fcayre committed:
ir.config.parameter groups_ids is gone in 11.0

So we use sudo instead to and write parameters

fcayre committed:
More trivial fixes: model and attribute names, string dates and the like

fcayre committed:
[fix] In 12.0, optin is to be performed when contract as 0 invoice

In v12 the contract invoice is created after the discount is computed,
so the optin must be performed when contract as 0 invoice, not 1.
This needs more thoughts as this is a fragile implementation, although a
test has been added.

fcayre committed:
Remove deprecates utility script to populate commown_cooperative_campaign

legalsylvain committed:
[REF] remove some modules from the CI, to avoid to install dependencies.
[REF] only ignore 2 modules, with failing tests.
[FIX] commown_cooperative_campaign: correct version.
[FIX] magic/python-magic installation. Use : external_dependencies_override
[FIX] test-requirements.txt : fix typo, and install 'account_payment_slimpay'
[REF]  urban_mine : set installable to False

fcayre committed:
Mute expected error logs in tests

fcayre committed:
Use untaxed amounts in tests to avoid l10n_generic_coa post init hook oddities

fcayre committed:
Be more defensive regarding cooperative campaign missing salt

It mays indeed crash during the validation of a sale, where contract end is set, calling TeleCommown WS.

fcayre committed:
Do not optout at a contract end if the discount was not actually valid

This fixes a bug where a customer has 2 contracts, one with a valid campaign,
the other without. Ending that last contract before the fix used to optout
the customer from the valid campaign of the other contract.

Also move optin to the discount line `compute` method so that it is not
called when checking coupon validity, for instance at the end of a contract.

fcayre committed:
Do not optin campaign in future invoice simulations

fcayre committed:
Port forgotten 10.0 cooperative campain coupon actions and test them

fcayre committed:
Port forgotten 10.0 cooperative campaign opt-in wizard and test it

fcayre committed:
Real instance does not like the model_coupon_coupon external id, so prefix with the right module

fcayre committed:
Fix crash when setting contract end date to False

Which happens in _compute_date_end of the contract module.

legalsylvain committed:
[META] enable black

legalsylvain committed:
[META] enable pyupgrade

legalsylvain committed:
[META] enable isort

legalsylvain committed:
[META] enable black (new version)

fcayre committed:
Fix name errors when using cooperative campaign late optin wizard

Add tests to cover this use case too.

github-grap-bot committed:
commown_cooperative_campaign 12.0.1.1.1

fcayre committed:
Add a model _description to silent a spurious warning

fcayre committed:
Remove the future invoices simulation feature

It is now replaced by the more general contract_forecast module.

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Add a small check to existing test + a comment

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Turn bypass_coop_campaigns context var into a dict

The idea here is to turn a True/ False global switch that avoid all cooperative
campaign WS calls alltogether into a description of discounts to be explicitely
accepted or rejected.

fcayre committed:
Factor out a cooperative campaign test utility class

... this class will be used in next commit.

fcayre committed:
Forecast cooperative campaign discounts as their last computation

To this end, we check for each applicable coop campaign discount if it was applied on
last invoice and keep that result for contract forecasts.

Note this last-invoice test is fragile for now as the data model does not cleanly store
which discounts were applied on each invoice line. This will be improved in an upcoming
commit.

fcayre committed:
Improve discount applied in invoice line test reliability

... by using the new invoice line <-> discount relations instead of
simple text matching on invoice line name.

fcayre committed:
Optimize variable discount computation a bit

Check validity dates before potentially a lot more complex discount conditions like
a http call for cooperative campaigns.

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Optimize forecast discount computation

... and more specifically the computation of the state of the cooperative campaign
discounts (like Telecommown2021): we only need to find the last invoiced line of a
contract line when it has a cooperative discount.

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Use __name__ as logger names, not __file__

This makes it easier to mute them in tests.
This is the standard Odoo way to name them and standardize it in our
code base.

fcayre committed:
Return a boolean value in last_invoice_cooperative_discount_state to avoid a warning when setting the context

Namely: 'Comparing apples and oranges: account.invoice.line() == True'

fcayre committed:
[tests] Fix spurious warnings in tests

... either by muting some loggers or by changing the test code slightly

fcayre committed:
[tests] Remove all test_queue_job_no_delay=True occurrences

... this is not a clean way to execute jobs and it generates a lot of warnings in tests.

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
[IMP] commown_cooperative_campaign: black, isort, prettier

fcayre committed:
[MIG] commown_cooperative_campaign: Migration to 16.0

fcayre committed:
[IMP] commown_cooperative_campaign: pre-commit execution

fcayre committed:
Add basic security for coupon late optin wizard

fcayre committed:
Adapt to contract_variable_discount model rename

fcayre committed:
Turn deprecated @job decorator into data in the db

fcayre committed:
[tests] Turn deprecated SavepointCase into TransactionCase

fcayre committed:
account.invoice has become account.move

... and its date been renamed from 'date_invoice' into 'date'.

fcayre committed:
[tests] Odoo exception 'name' attribute is now deprecated

fcayre committed:
[tests] Turn setUp into setUpClass to speed-up tests

fcayre committed:
[tests] Add a test to improve coverage

One of the previous test was partially wrong and was fixed at the same time
(they share data with the new one).

fcayre committed:
Simplify code that displays user subscription status vs. a cooperative discount

The 'important events' service cannot return more than one subscription when
called for a single customer.

This makes it possible to simplify code that does not make this assumption.

This should also mechanically improve test coverage.

fcayre committed:
Remove useless string concatenation

... which probably came from black joining string lines that had become short enough.

fcayre committed:
Add trivial tests to improve coverage

Honeyxilia commented on commown_cooperative_campaign/__manifest__.py:
```diff
@@ -15,6 +15,7 @@
         "queue_job",
```
Espace manquant dans le message de commit c:

Honeyxilia commented on commown_cooperative_campaign/tests/test_coupon.py:
```diff
@@ -219,6 +220,62 @@ def test_action_optin_status_with_passed_unsubscription(self):
+    def test_coop_partner_identifier_ok(self):
+        self.assertEqual(
+            self.campaign.coop_partner_identifier(self.partner),
+            "04ca4c637f7445ab54673d8923a95f847a2c97f8a52ed2c8bd2b5381f666bfb7",
```
une valeur en dur comme cela me semble assez étrange - est-ce que c'est bien fiable ?
d'où vient cet identifiant (en dehors du champ dans la campagne) et est-il obtenable aisément ?

9184Lucky replied:
Il est fait par un hash avec le numéro de téléphone... Peut être possible de le refaire là pour ne pas sortir cette valeur du chapeau mais vu que on fixe les valeur du country_code et du numéro de teléphone dans le setup on est sur que d'obtenir toujours la même valeur !

Honeyxilia replied:
Dans ce cas, je pense que pour la clarté, il serait mieux de réinsérer l'opération de calcul de hash ici, mais si c'est fiable c'est okie c:

fcayre replied:
oui c'est facilement reproductible, ça vient du sel de chiffrement des identifiants clients pour qu'ils soient inutilisables en cas de vol de la base de données Telecommown. Reproduire la formule qui permet de le retrouver dans le test au lieu d'utiliser la valeur obtenue en dur reviendrait à copier-coller le code testé dans les tests, ce qui est absurde.

Honeyxilia replied:
okiee

fcayre committed:
Remove useless None default value for dictionary get

None is already the default value.

fcayre committed:
Test cooperative campaign error case and fix double error logging

fcayre committed:
Add test to improve coverage

fcayre committed:
Simplify super calls

fcayre committed:
Remove useless code

The check that the campaign is cooperative is already made
in the customer key computation, which returns nothing when
the campaign is not cooperative. This is enough to make sure
the opt-out service will not be called and this is already
tested.

fcayre committed:
[tests] Add a docstring

fcayre committed:
Refactor a bit double-optin handling to remove code

The 422 http error code was already handled in the coop_ws_optin
function call. The only missing part was the error logging to ease
debugging in production. So this part of the code was moved and
everything else removed to avoid duplication.

fcayre commented:
/ocabot merge patch
